src/animated/Controller.ts

```diff
@@ -190,310 +198,322 @@ class Controller<DS extends object = any> {
-    } else {
-      this._diff(props)._start(onEnd)
+  private _run(props: UpdateProps<State>, onEnd?: OnEnd) {
+    if (this._diff(props)) {
+      this._animate(this.props)
+      if (is.arr(props.to) || is.fun(props.to)) {
+        this._runAsync(props, onEnd)
+      } else {
+        this._start(onEnd)
+      }
     }
   }
 
   // Start an async chain or an async script.
-  private _runAsync(guid: number, props: UpdateProps<DS>, onEnd?: OnEnd) {
-    let queue = Promise.resolve()
+  private _runAsync({ to }: UpdateProps<State>, onEnd?: OnEnd) {
+    const { animations } = this
+
+    // The `stop` and `destroy` methods clear the animation map.
+    const isCancelled = () => animations !== this.animations
+
+    let last: Promise<void>
+    const next = (asyncProps: SpringProps<State>) => {
+      if (isCancelled()) throw this
+      if (to !== this.props.to) return
+      return (last = new Promise<any>(done => {
+        this.update(asyncProps).start(done)
+      }))
+    }
 
-    const { to } = props
+    let queue = Promise.resolve()
     if (is.arr(to)) {
-      to.forEach((p, i) => {
-        queue = queue.then(() => {
-          if (guid !== this.guid) return
-          const fresh = { ...props, ...interpolateTo(p) }
-          if (is.arr(fresh.config)) fresh.config = fresh.config[i]
-          return new Promise<any>(r => this._diff(fresh)._start(r))
-        })
-      })
+      to.forEach(props => (queue = queue.then(() => next(props))))
     } else if (is.fun(to)) {
-      // Throw this to interrupt an infinite script
-      const stopToken = {}
-
-      let i = 0
-      let last: Promise<any>
-      queue = queue.then(() =>
-        to(
-          // next(props)
-          p => {
-            if (guid !== this.guid) throw stopToken
-            const fresh = { ...props, ...interpolateTo(p) }
-            if (is.arr(fresh.config)) fresh.config = fresh.config[i++]
-            return (last = new Promise(r => this._diff(fresh)._start(r)))
-          },
-          // stop(finished)
-          this.stop.bind(this)
-        ).then(
-          () => last,
-          err => {
-            if (err !== stopToken) throw err
-          }
-        )
-      )
+      queue = queue.then(() => to(next, this.stop.bind(this)).then(() => last))
     }
 
+    queue = queue.catch(err => {
+      // "this" is thrown when the animation is cancelled
+      if (err !== this) console.error(err)
+    })
+
     if (onEnd) {
-      queue.then(() => onEnd(guid === this.guid))
+      queue.then(() => onEnd(!isCancelled()))
     }
   }
 
-  private _diff(props: UpdateProps<DS>) {
-    Object.assign(this.props, props)
-    let {
-      from = {} as any,
-      to = {} as any,
-      config = {},
-      reverse,
-      attach,
-      reset,
-      immediate,
-      onStart,
-    } = this.props
+  // Merge every fresh prop. Return false if no props were fresh.
+  private _diff({ timestamp, ...props }: UpdateProps<State>): boolean {
```

Comment:
Now, the `_diff` method only updates `this.props` and `this.timestamps`, while the new `_animate` method updates `this.animations`.